lib/http_get.js

```diff
@@ -0,0 +1,312 @@
+            opContext.log.error(
+                'Failure to persist fragments to repair', err);
+            // Return persistence error
+            // TODO: not sure this a proper HTTP error....
+            // TODO: must be a better way than disabling it
+            /* eslint-disable no-param-reassign */
+            err.infos = { status: 500, method: 'GET' };
+            opContext.failedToPersist = true;
+            /* eslint-enable no-param-reassign */
+            return callClientCb ? callback(err) : null;
+        }
+
+        /* Error persistence OK, return already encountered
+         * error (might be null) */
+        return callClientCb ? callback(currentError) : null;
+    });
+}
+
+function _handleCorruption(reqContext, corruptError, errorAgent) {
+    const toRepair = {
+        rawKey: reqContext.opContext.rawKey,
+        fragments: [[reqContext.chunkId, reqContext.fragmentId]],
+    };
+    _handleErrors(reqContext.opContext, errorAgent, toRepair,
+                  null, corruptError, false);
+}
+
+function getCorruptionCheckedStream(httpReply, reqContext, errorAgent) {
+    const ctypes = protocol.helpers.parseReturnedContentType(
+        httpReply.headers['content-type']);
+    const size = ctypes.get('data');
+    const expectedCRC = ctypes.get('$crc.data');
+
+    if (!size || !expectedCRC) {
+        return httpReply;
+    }
+
+    let bytesUntilCRC = size;
+    let endDataBuffer = null;
+    let crcBuffer = null;
+    const corruptSaferStream = new stream.Transform({
+        transform(chunk, encoding, callback) {
+            if (chunk.length < bytesUntilCRC) {
+                this.push(chunk);
+                bytesUntilCRC -= chunk.length;
+                callback();
+                return;
+            }
+
+            endDataBuffer = endDataBuffer === null ?
+                chunk.slice(0, bytesUntilCRC) :
+                Buffer.concat([endDataBuffer, chunk.slice(0, bytesUntilCRC)]);
+
+            crcBuffer = crcBuffer === null ?
+                chunk.slice(bytesUntilCRC) :
+                Buffer.concat([crcBuffer, chunk.slice(bytesUntilCRC)]);
+
+            /* Partial CRC */
+            if (crcBuffer.length < 12) {
+                bytesUntilCRC -= chunk.length;
+                callback();
+                return;
+            }
+
+            /* Check actual CRC
+             * First 4 bytes are data crc */
+            const actualCRC = crcBuffer.slice(0, 4).readUInt32LE();
+            if (actualCRC !== expectedCRC) {
+                reqContext.opContext.log.error(
+                    'Corrupted data',
+                    { expectedCRC, actualCRC });
+                const corrupted = new Error('Corrupted data');
+                corrupted.infos = {
+                    status: 422,
+                    method: 'GET',
+                };
+                this.emit('error', corrupted);
+                _handleCorruption(reqContext, corrupted, errorAgent);
+                callback();
```

Comment:
So this stream is never calling its callback with an error, even in case of pieces-corrupted ?  I guess that's why it's a corrupt-safe Rstream ?

Reply:
I may have missed part of the documentation on Transform here. The callback just asks it continue, but I have emitted an error manually 2 lines above. Perhaps I just have to use callback(corrupted) to have the correct behaviour.
Normally we delay the sending of the last chunk of data until we received and check the CRCs. So we are probably 100% safe. I kept safer because uncertainty is mostly limited to my JS (in)experience :S